ddtrace/profiling/collector/_traceback.pyx

```diff
@@ -49,11 +102,25 @@ cpdef pyframe_to_frames(frame, max_nframes):
     :return: The serialized frames and the number of frames present in the original traceback."""
     frames = []
     nframes = 0
-    while frame is not None:
-        nframes += 1
-        if len(frames) < max_nframes:
-            code = frame.f_code
-            lineno = 0 if frame.f_lineno is None else frame.f_lineno
-            frames.append((code.co_filename, lineno, code.co_name, _extract_class_name(frame)))
-        frame = frame.f_back
+    # Python 3.11 moved PyFrameObject members to internal C API and cannot be directly accessed
+    IF PY_MAJOR_VERSION > 3 or (PY_MAJOR_VERSION == 3 and PY_MINOR_VERSION >= 11):
+        cdef PyFrameObject* pyframe = <PyFrameObject*> frame
+        while pyframe != NULL:
+            nframes += 1
+            if len(frames) < max_nframes:
+                code = PyFrame_GetCode(pyframe)
+                lineno = PyFrame_GetLineNumber(pyframe)
+                lineno = 0 if lineno is None else lineno
+                frames.append(((<object>code).co_filename, lineno, (<object>code).co_name, _extract_class_name(<object>pyframe)))
+                Py_XDECREF(<PyObject*>code)
+            pyframe = PyFrame_GetBack(pyframe)
+            # FIXME: Where to Py_XDECREF(cframe)?
```

Comment:
`s/cframe/pyframe/` ?

Reply:
Thanks for the catch!

I am looking for some suggestions on where to call `Py_XDECREF(pyframe)` though. Since `PyFrame_GetBack(pyframe)` returns a strong reference of the `PyFrameObject*`, should I be decrementing the refcount for every time we call `PyFrame_GetBack(pyframe)`? But since we're using the result of the `PyFrame_GetBack()` in the while loop, I don't think I can decrement the refcount without interfering with the rest of the function.

Reply:
`Py_XDECREF(pyframe)` has to be called every time you are done with that particular `pyframe` reference, so within the loop. The cast to `<object>` is taking care of ref counting when passing values around in Cython, so as long as our ref counting for when we invoke the C API directly is correct we should be good.